### ISSUE TITLE
Add tactic to remove typeclass instances

### ISSUE THREAD

**Issue by MSoegtropIMC** [coq/coq#7342]
I am following up on the discussion on coq-club with @mattam82 with subject "Removing type class instances".

The following sequence:

    Require Import Morphisms.
    Require Import Classes.Morphisms_Prop.
    Remove Hints all_iff_morphism : typeclass_instances.
    Print Instances Proper.

still shows the all_iff_morphism instance. According to Matthieu it is not used, though.

Matthieu suggested to create a dedicated tactic to remove type classes, but fixing Remove Hints would of cause also work.

Btw.: Remove Hints <instance> : typeclass_instances is not documented.

Best regards,

Michael

**Comment by ppedrot** [comment]
Yes, there is no synchronization between the hint database and the typeclass database. Rather than providing another command, I think we should work on that...

**Comment by maximedenes** [comment]
I was a bit puzzled by the claim that the typeclass database is useful only for the `Print Instances` command. If that is true, why not just add one bit to the entries of the hint database, instead of having two databases?

**Comment by ppedrot** [comment]
Technically we could even just look for hints that have a pattern corresponding to the class in the hint base...

**Comment by Zimmi48** [comment]
@mattam82 did not propose a new tactic but a new command.
Also, we talked IRL together about this and concluded that the second instance database should just be removed and the `Print Instances` command should use what it finds in the hint base, like @ppedrot just suggested.

**Comment by Zimmi48** [comment]
BTW `Remove Hints` is documented here https://coq.inria.fr/refman/proof-engine/tactics.html#coq:cmd.remove-hints
Do you mean there should be a link to it in the typeclass chapter?

**Comment by MSoegtropIMC** [comment]
> Do you mean there should be a link to it in the typeclass chapter?

Yes. Since there is a separate command for creating instances, it would make sense to describe nearby how they are removed.

And sorry with being sloppy about tactics and commands. Of cause I meant commands.

**Comment by Zimmi48** [comment]
Removing the second instance database is blocked not just on plugging `Print Instances` onto `Print HintDb` but also on having a general discharging mechanism for hints (currently the discharge mechanism for this special instance database takes care of putting back the instances in the hint database `typeclass_instances`).

cc @herbelin
